crates/primitives/src/transaction/mod.rs

```diff
@@ -620,6 +620,75 @@ impl TransactionSigned {
         initial_tx.hash = initial_tx.recalculate_hash();
         initial_tx
     }
+
+    /// Decodes a eip2718 transaction where tx is [id, rlp(typedtx)]
+    pub fn decode_enveloped(&self, tx: Bytes) -> Result<Self, DecodeError> {
+        let mut data = tx.as_ref();
+
+        let first_header = Header::decode(&mut data)?;
+
+        if data[0] > 0x7f {
+            // legacy transaction
+            let mut transaction = Transaction::Legacy(TxLegacy {
+                nonce: Decodable::decode(&mut data)?,
+                gas_price: Decodable::decode(&mut data)?,
+                gas_limit: Decodable::decode(&mut data)?,
+                to: Decodable::decode(&mut data)?,
+                value: Decodable::decode(&mut data)?,
+                input: Bytes(Decodable::decode(&mut data)?),
+                chain_id: None,
+            });
+            let (signature, extracted_id) = Signature::decode_with_eip155_chain_id(&mut data)?;
+            if let Some(id) = extracted_id {
+                transaction.set_chain_id(id);
+            }
+
+            let tx_length = first_header.payload_length + first_header.length();
+            let hash = keccak256(&data[..tx_length]);
+            let signed = TransactionSigned { transaction, hash, signature };
+            Ok(signed)
+        } else {
+            let tx_type = *data.first().ok_or(DecodeError::InputTooShort)?;
+            data.advance(1);
+            // decode the list header for the rest of the transaction
+            let header = Header::decode(&mut data)?;
+            if !header.list {
+                return Err(DecodeError::Custom("typed tx fields must be encoded as a list"))
+            }
+
```

Comment:
same here.

please unify as much as possible.

Reply:
I'm sorry but I don't  think understand clearly what you mean by unify.

Reply:
I think the general idea is that if there is code duplication between this method and the `Decodable` implementation, then that code should be abstracted into a method and used in both this method + the `Decodable` implementation